src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
+    //   shuffle, the LIS optimization becomes less helpful, so it seems not
+    //   worth the code at this point, but could consider if a compelling case
+    //   arises.
+    while (oldStartIndex <= oldEndIndex && newStartIndex <= newEndIndex) {
+      if (oldStartPart == null) {
+        // Old part at head has already been used; skip
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (oldEndPart == null) {
+        // Old part at tail has already been used; skip
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newStartResult.key) {
+        // Old head matches new head; update in place
+        newParts[newStartIndex] = updatePart(oldStartPart, newStartResult);
+        oldStartPart = oldParts[++oldStartIndex];
+        newStartResult = newResults[++newStartIndex];
+      } else if (oldEndPart.key == newEndResult.key) {
+        // Old tail matches new tail; update in place
+        newParts[newEndIndex] = updatePart(oldEndPart, newEndResult);
+        oldEndPart = oldParts[--oldEndIndex];
+        newEndResult = newResults[--newEndIndex];
+      } else if (!newKeyToIndexMap.has(oldStartPart.key)) {
+        // Old head is no longer in new list; remove
+        removePart(oldStartPart);
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (!newKeyToIndexMap.has(oldEndPart.key)) {
+        // Old tail is no longer in new list; remove
+        removePart(oldEndPart);
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newEndResult.key) {
+        // Old head matches new tail; update and move to new tail
+        newParts[newEndIndex] = updatePart(oldStartPart, newEndResult);
+        movePart(directivePart, oldStartPart, newParts[newEndIndex + 1]);
+        oldStartPart = oldParts[++oldStartIndex];
+        newEndResult = newResults[--newEndIndex];
+      } else if (oldEndPart.key == newStartResult.key) {
+        // Old tail matches new head; update and move to new head
+        newParts[newStartIndex] = updatePart(oldEndPart, newStartResult);
+        movePart(directivePart, oldEndPart, oldStartPart);
+        oldEndPart = oldParts[--oldEndIndex];
+        newStartResult = newResults[++newStartIndex];
+      } else {
+        // Any mismatches at this point are due to additions or moves; see if
+        // we have an old part we can reuse and move into place
+        const oldIndex = oldKeyToIndexMap.get(newEndResult.key);
+        const oldPart = oldIndex !== undefined ? oldParts[oldIndex] : null;
+        if (oldPart == null) {
+          // No old part for this result; create a new one and insert it
+          let newPart = createPart(
+              directivePart, newEndResult, newParts[newEndIndex + 1]);
+          newParts[newEndIndex] = newPart;
+        } else {
+          // Reuse old part
+          newParts[newEndIndex] = updatePart(oldPart, newEndResult);
+          movePart(directivePart, oldPart, newParts[newEndIndex + 1]);
```

Comment:
Again, clearer and equivalent to move `newParts[newEndIndex]`

Reply:
As discussed, `newParts` fills in outside in, so `newParts[newEndIndex]` is not a valid reference yet.